src/_guides/language/evolution.md

```diff
@@ -127,25 +151,99 @@ void main() {
 }
 ```
 
-## Dart 2.8
+### Dart 2.8
 
 Dart 2.8 didn't add any features to the Dart language, but it did
 contain a number of preparatory [breaking changes][2.8 breaking changes] to ensure great
 nullability-related usability and performance in the upcoming
-null safety feature.
+[null safety][] feature.
 
 It also contained a faster pub tool, and a new [pub outdated][] command.
 
-## Dart 2.9
+### Dart 2.9
 
 Dart 2.9 didn't add any features to the Dart language.
 
-## Dart 2.10
+### Dart 2.10
 
 Dart 2.10 didn't add any features to the Dart language,
 but it added an expanded [`dart` tool][dart-tool] that's 
 analogous to the Flutter SDK's `flutter` tool.
 
+## Language versioning
+
+Before language versioning,
+Dart tools assumed that all Dart code could use all features
+that were supported by the tool's SDK version.
+This assumption causes problems when a package
+that uses a _newer_ language feature
+has [SDK constraints][] that are too low.
+For example, consider a package that uses set literals (2.2)
+but has the SDK constraints `">=2.0.0 <3.0.0"`.
+People using Dart 2.0.0 or 2.1.0 can download the package,
+but their tools can't understand the set literals.
```

Comment:
I believe even before language versioning, our tools would warn if you used a language feature that was added to Dart after your SDK's minimum version. That's a fairly straightforward set of warnings we could support whenever if we wanted to.

Language versioning is quite a bit more complex. It's really about being able to handle the fact that the *same* Dart code may mean *different* things in different versions of Dart. Concretely, in 2.10:

```dart
int i;
```

Declares a variable whose value can be `null`, and has no errors. Once null safety ships, that same program now has a compile error.

The example you describe with set literals is one of *forwards* incompatibility: an old SDK can't handle new set literals. That's definitely an issue, but language versioning is more about handling *backwards* incompatibility. It lets us change the language in incompatible ways without breaking existing code.